### PR TITLE
Mdoc incorrect signature for return types

### PR DESCRIPTION
**Pull request by asokolovskiy15**

### CODE REVIEW

**Comment by asokolovskiy15** [comment]
@joelmartinez 

It's a temporary commit. Here, we removed namespace when return type is generic and symbols '&' and '@'. Also, we added Ref type attribute. Could you please take a look at the commit, and tell if we've done correct changes.
I will show a few examples of what has changed
```
 -  <ReturnValue>
 -   <ReturnType>System.Int32@</ReturnType>
 -  </ReturnValue>
 +  <ReturnValue RefType="Ref">
 +   <ReturnType>System.Int32</ReturnType>
 +  </ReturnValue>
```

```
 -  <ReturnValue>
 -   <ReturnType>System.Int32&amp</ReturnType>
 -  </ReturnValue>
 +  <ReturnValue RefType="Readonly">
 +   <ReturnType>System.Int32</ReturnType>
 +  </ReturnValue>
```

Also, our changes affect not only `ReturnValue` node.
This is the `Parameters` node after changes

```
       <Parameters>
-        <Parameter Name="t" Type="T&amp;" RefType="ref" />
-        <Parameter Name="u" Type="U&amp;" RefType="ref" />
+        <Parameter Name="t" Type="T" RefType="ref" />
+        <Parameter Name="u" Type="U" RefType="ref" />
       </Parameters>
```

Can I continue to edit integration tests?

**Comment by joelmartinez** [comment]
@asokolovskiy15 yes ... this looks good. Please include in at least one test, running mdoc multiple times. I want to make sure that after this change, the reflected member info will be correctly matched up with the existing XML docs :)

**Comment by asokolovskiy15** [comment]
@joelmartinez I've updated current commit. But it's temporary still. Could you please pay attention to the test `check-monodocer-importecmadoc`

After changes, one of the integration tests was failed. Before changes,  `TestEcmaDocs.xml` didn't import documentation generated in test due to difference in signatures. 
Signature that was before:
`<Parameter Name="array" Type="T[]&amp"/>`

Current signature
`<Parameter Name="array" Type="T[]"/>`

`TestEcmaDocs.xml` has following signature:

`<Parameter Name="array" Type="T[]"/>`

As you can see current signature matches with signature in `TestEcmaDocs.xml`, We think that it is a reason why after the fix documentation for a node is imported

We think that test has a correct result now and we would like that you'll double check it

**Comment by asokolovskiy15** [comment]
>should the current Parameter node have a RefType attribute?

The parameters can contain `ref` modifier, so `RefType` attribute seems to be required here.

Our changes in `AppendRefTypeName` method caused changes in generation of `SlashDocMemberFormatter` signature.

For method `M1` from `Widget` class, `SlashDocMemberFormatter` generates:

```
 M:Mono.DocTest.Widget.M1(System.Char,System.Signle,Mono.DocTest.DocValueType)
```
It causes the issue when we run the following command: 
```
 update -i DocTest.xml -o en.actual DocTest.dll
```

Class `MsxdocDocumentationImporter` tries to find matching between `Doc-Test.xml` and `Doc-Test.dll` elements, and can't do it because 
```
M:Mono.DocTest.Widget.M1(System.Char,System.Signle@,Mono.DocTest.DocValueType@)
!=
M:Mono.DocTest.Widget.M1(System.Char,System.Signle,Mono.DocTest.DocValueType)
```

We found the code where this comparison happens:
```C#
return slashdocs.SelectSingleNode ("doc/members/member[@name='" + slashdocsig + "']");
```

Should we fix this logic to resolve this situation?

Also, we've updated the pull request, where few integration were edited. Only tests where the mentioned situation happens still fail.

**Comment by joelmartinez** [comment]
Ahh, I see ... so in this case, we're going to have to add some sort of flag, or way to determine the usage ... because in this case when we're trying to generate the `M:...` form of the signature, we'll need to _include_ the `@`, as that's an important part of the signature. Whereas, when we're storing it in the XML, we want to separate that information out into more "strongly typed" metadata.

**Comment by asokolovskiy15** [comment]
@joelmartinez 
We fixed this issue: `MsxdocSlashDocMemberFormatter` was added. It generates the same signature as an imported xml file contains.

File: mdoc/Makefile [hunk 1]
```diff
@@ -709,7 +717,8 @@ check-doc-tools: 	check-monodocer-since \
 	check-monodocer-cppwinrtUwp\
 	check-monodocer-frameworkalternate-fromold \
 	check-monodocer-frameworkalternate \
-	check-monodocer-embedded-type \
+	check-monodocer-embedded-type \
+	check-ignored-namespace-generic \
```
Comment (joelmartinez) [inline mdoc/Makefile]: I mentioned in [this comment](https://github.com/mono/api-doc-tools/pull/251#issuecomment-388083113) that we should have a test here to test the efficacy of importing some of these elements, who's `Type` parameter has now changed (for example, a `ref T` parameter .. I didn't see any of those in `TestEcmaDocs.xml`).

The DocXml that is generated by the compiler should have the triple slash docs in the code (so you can verify that it makes it over into the EcmaXML correctly).

File: mdoc/Test/msxdoc-expected.importslashdoc.xml [hunk 2]
```diff
@@ -729,7 +729,7 @@ class Example {
             <remarks>
                 <c>M:Mono.DocTest.Widget.M0</c>.</remarks>
         </member>
-        <member name="M:Mono.DocTest.Widget.M1(System.Char,System.Single@,Mono.DocTest.DocValueType@)">
```
[the remark this answers is not in the document]
Comment (joelmartinez) [reply mdoc/Test/msxdoc-expected.importslashdoc.xml]: @asokolovskiy15 yes I understand why you changed the file :P however, the issue is that it's an incorrect output. The purpose of the `export-msxdoc` subcommand is to produce a file that would be equivalent to the output of `/doc` from the compiler. So to use an explicit example:

1. with dotnet core installed locally
     - `mkdir csdoccomment`
     - `cd csdoccomment`
     - `dotnet new library`
1. edit the `.csproj` file to add the `DocumentationFile` element
    ```xml
    <Project Sdk="Microsoft.NET.Sdk">
      <PropertyGroup>
        <TargetFramework>netstandard2.0</TargetFramework>       
        <DocumentationFile>bin\$(Configuration)\$(TargetFramework)\$(AssemblyName).xml</DocumentationFile>
      </PropertyGroup>
    </Project>
    ```
1. Add a member that reflects the particular example we're discussing here:
    ```csharp
    using System;
    
    namespace csdoccomment
    {
        public class Class1
        {
            /// <summary>test</summary>
            public void M1 (char c, out float f, ref byte v)
            {
                f = 1;
            }
        }
    }
    ```
1. `dotnet build`

Once you do this, you'll find `csdoccomment.xml` in the bin folder, and it looks like this

```xml
<?xml version="1.0"?>
<doc>
    <assembly>
        <name>csdoccomment</name>
    </assembly>
    <members>
        <member name="M:csdoccomment.Class1.M1(System.Char,System.Single@,System.Byte@)">
            <summary>test</summary>
        </member>
    </members>
</doc>

```

So as you can see, `M:csdoccomment.Class1.M1(System.Char,System.Single@,System.Byte@)` **contains** the `@` characters ... thus, the output of the `export-msxdoc` must match this. the formatter used there needs to take this into account, and watch for the `RefType="ref"` attribute on the node, and generate accordingly.

**Comment by asokolovskiy15** [comment]
@joelmartinez 
I've updated the pull request. Integration test `check-multiple-mdoc` was added. As you suggested, we take `RefType` into account when generate argument type.

**Comment by joelmartinez** [comment]
build

**Pull request merged**